### PR TITLE
MON-3231: add metric for the configured profile

### PR DESCRIPTION
**Pull request by rexagod**
- adds a metric with the label value set to the current collection profile.

* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: pkg/metrics/metrics.go
Comment (simonpasquier) [inline pkg/metrics/metrics.go:27]: Can you expand a bit? I'm not sure to know to which component "once GA'd" applies :)

**Review by simonpasquier — COMMENTED** [review]
Since collection profiles are gated via TechPreview, I would only expose the metric if tech preview is explicitly enabled.

I'd also recommend exposing all collection profile values, e.g.

```
cluster_monitoring_operator_collection_profile{..., collection_profile="full"} 1
cluster_monitoring_operator_collection_profile{..., collection_profile="minimal"} 0
```

Having 0 for all series would mean that the collection profile is misconfigured.

File: pkg/manifests/config.go
Comment (jan--f) [inline pkg/manifests/config.go:220]: Iiuc this is a no-op, since `CollectionProfile` will default to `full`: https://github.com/openshift/cluster-monitoring-operator/blob/ec2e0d3da7199f68c73f5a081ebf2a131d56f243/pkg/manifests/config.go#L303

**Review by jan--f — COMMENTED** [review]
Generally this looks good to me. Can you please add separate commits for the lint fix, the metrics refactor and for adding the new metric? This will make future change archeology much easier.
Also this does not add the telemetry config yet, we could add the CMO side for this here too?

One thing I'd like to raise for discussion: Should we maybe add a `cluster_monitoring_operator_configuration` metric instead of single metrics for config options we're interested in? We're mostly interested in the label we set anyway and I think such a metadata metric can make it easier to add additional signals and send them to telemetry.

**Comment by jan--f** [comment]
> One thing I'd like to raise for discussion: Should we maybe add a cluster_monitoring_operator_configuration metric instead of single metrics for config options we're interested in? We're mostly interested in the label we set anyway and I think such a metadata metric can make it easier to add additional signals and send them to telemetry.

Briefly discussed in [slack](https://redhat-internal.slack.com/archives/G79AW9Q7R/p1688373468751679), individual metrics have clear benefits.

File: pkg/metrics/metrics.go
Comment (rexagod) [reply pkg/metrics/metrics.go]: ACK, updated.

File: pkg/manifests/config.go
Comment (rexagod) [reply pkg/manifests/config.go]: Ah, right!

**Comment by rexagod** [comment]
> Can you please add separate commits for the lint fix, the metrics refactor and for adding the new metric? This will make future change archeology much easier.

@jan--f Does it sound good if do this once the PR is LGTM'd and good to merge? That way I can keep adding any lint fixes or refactors during further reviews and split them all in the end.

File: Documentation/data-collection.md
Comment (simonpasquier) [inline Documentation/data-collection.md:593]: You want to aggregate and record the metric to remove variable/unnecessary labels such as pod, job, service and instance. E.g. `max by(profile) (cluster_monitoring_operator_collection_profile)`

https://rhobs-handbook.netlify.app/products/openshiftmonitoring/telemetry.md/#configure-recording-rules

File: pkg/manifests/config.go
Comment (simonpasquier) [inline pkg/manifests/config.go:231]: (nit)
```suggestion
			var v float64
```

File: pkg/manifests/config.go
Comment (simonpasquier) [inline pkg/manifests/config.go:237]: (nit) we can get rid of the intermediate variable
```suggestion
		if !slices.Contains(SupportedCollectionProfiles, c.ClusterMonitoringConfiguration.PrometheusK8sConfig.CollectionProfile) {
```

File: pkg/metrics/metrics.go
Comment (simonpasquier) [inline pkg/metrics/metrics.go:31]:
```suggestion
}, []string{"profile"})
```

File: pkg/metrics/metrics.go
Comment (simonpasquier) [inline pkg/metrics/metrics.go:30]: I don't think that the dimension of the metric is unbounded.

File: pkg/metrics/metrics.go
Comment (simonpasquier) [inline pkg/metrics/metrics.go:26]:
```suggestion
	Name:           "cluster_monitoring_operator_collection_profile",
```

File: pkg/metrics/metrics.go
Comment (simonpasquier) [inline pkg/metrics/metrics.go:27]:
```suggestion
	Help:           "Information about collection profiles. Set to 1 for the configured collection profile, else 0.",
```

File: pkg/manifests/config.go
Comment (simonpasquier) [inline pkg/manifests/config.go:35]: (nit) do we need the import alias?

File: Documentation/data-collection.md
Comment (rexagod) [reply Documentation/data-collection.md:593]: Ah! Apologies, I totally missed that. Addressed.

File: pkg/metrics/metrics.go
Comment (rexagod) [reply pkg/metrics/metrics.go:30]: ACK, I thought about using this in a best practice sort of way, but it doesn't really help here, since it's guaranteed the cardinality will be bounded.

Comment (rexagod) [reply pkg/metrics/metrics.go:30]: I've removed the comments.

File: pkg/metrics/metrics.go
Comment (jan--f) [inline pkg/metrics/metrics.go:24]: This comment needs an update to be consistent with the metric help.

File: pkg/metrics/metrics.go
Comment (jan--f) [inline pkg/metrics/metrics.go:25]:
```suggestion
var CollectionProfile = metrics.NewGaugeVec(&metrics.GaugeOpts{
```
And update the comment above?

**Review by jan--f — CHANGES_REQUESTED** [review]

**Comment by jan--f** [comment]
/lgtm
/hold
for better commits :)

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (simonpasquier) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:582]:
```suggestion
    # collection profile.

File: jsonnet/rules.libsonnet
Comment (simonpasquier) [inline jsonnet/rules.libsonnet:335]: We only need to report the selected profile.
```suggestion
          expr: 'max by (profile) (cluster_monitoring_operator_collection_profile == 1)',
```

**Review by simonpasquier — COMMENTED** [review]
one last nit :)

File: pkg/metrics/metrics.go
Comment (rexagod) [inline pkg/metrics/metrics.go:29]: This entire file came up as one single blob while [patching](https://github.com/tpope/vim-fugitive/blob/43f18ab9155c853a84ded560c6104e6300ad41da/doc/fugitive.txt#L313), and although this snippet comes under the metric addition, the file this is in, is itself a part of the refactor, so I included this here. PLMK if this doesn't make sense and I'll move this bit over to the latest commit.

Comment (rexagod) [reply pkg/metrics/metrics.go:29]: Let me break this from here and move it into the latest commit, it makes more sense there, I think.

**Pull request merged**
